workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/flows/RedeemToken.kt

```diff
@@ -1,85 +1,172 @@
+        @Suspendable
+        override fun call(): SignedTransaction {
+            // Receive a redeem notification from the party. It tells us if we need to sign up for token updates or
+            // generate a confidential identity.
+            val redeemNotification = otherSession.receive<TokenRedeemNotification<T>>().unwrap { it }
+
+            // Request confidential identity, if necessary.
+            val otherIdentity = if (redeemNotification.anonymous) {
+                subFlow(RequestConfidentialIdentity.Initiator(otherSession)).party.anonymise()
+            } else otherSession.counterparty
+
+            val stateAndRefsToRedeem = subFlow(ReceiveStateAndRefFlow<AbstractToken>(otherSession))
+            check(stateAndRefsToRedeem.isNotEmpty()) {
+                "Received empty list of states to redeem."
+            }
+            checkSameIssuer(stateAndRefsToRedeem, ourIdentity)
+            checkSameNotary(stateAndRefsToRedeem)
+            checkOwner(serviceHub.identityService, stateAndRefsToRedeem)
+            val notary = stateAndRefsToRedeem.first().state.notary
+            val txBuilder = TransactionBuilder(notary = notary)
+            if (redeemNotification.amount == null) {
+                generateExitNonFungible(txBuilder, stateAndRefsToRedeem.first() as StateAndRef<NonFungibleToken<T>>)
+            } else {
+                TokenSelection(serviceHub).generateExit(txBuilder, stateAndRefsToRedeem as List<StateAndRef<FungibleToken<T>>>, redeemNotification.amount, otherIdentity)
+            }
+            // TODO Does it make sense for the issuer to use confidential identities?
+            val partialStx = serviceHub.signInitialTransaction(txBuilder, ourIdentity.owningKey)
+            val stx = subFlow(CollectSignaturesFlow(partialStx, listOf(otherSession)))
+            return subFlow(FinalityFlow(transaction = stx, sessions = listOf(otherSession)))
+        }
+    }
+
+    // Check that all states share the same notary.
+    @Suspendable
+    private fun checkSameNotary(stateAndRefs: List<StateAndRef<AbstractToken>>) {
+        val notary = stateAndRefs.first().state.notary
+        check(stateAndRefs.all { it.state.notary == notary }) {
+            "All states should have the same notary. Automatic notary change isn't supported for now."
+        }
+    }
+
+    // Checks if all states have the same issuer. If the issuer is provided as a parameter then it checks if all states
+    // were issued by this issuer.
+    @Suspendable
+    private fun checkSameIssuer(stateAndRefs: List<StateAndRef<AbstractToken>>, issuer: Party? = null) {
+        val issuerToCheck = issuer ?: stateAndRefs.first().state.data.issuer
+        check(stateAndRefs.all { it.state.data.issuer == issuerToCheck }) {
+            "Tokens with different issuers."
+        }
+    }
+
+    // Check if owner of the states is well known. Check if states come from the same owner.
```

Comment:
they should also only come from the node which is sending them.

Reply:
good spot!